src/coreclr/System.Private.CoreLib/src/System/Runtime/ControlledExecution.CoreCLR.cs

```diff
@@ -0,0 +1,173 @@
+            private Thread? _thread;
+
+            public Execution(Action action, CancellationToken cancellationToken)
+            {
+                _action = action;
+                _cancellationToken = cancellationToken;
+            }
+
+            public void Run()
+            {
+                Debug.Assert((_state & (int)State.Started) == 0 && _thread == null);
+
+                // Recursive ControlledExecution.Run calls are not supported
+                if (t_execution != null)
+                    throw new InvalidOperationException(SR.InvalidOperation_NestedControlledExecutionRun);
+
+                _thread = Thread.CurrentThread;
+
+                try
+                {
+                    try
+                    {
+                        // As soon as the Started flag is set, this thread may be aborted asynchronously
+                        if (Interlocked.CompareExchange(ref _state, (int)State.Started, (int)State.None) == (int)State.None)
+                        {
+                            t_execution = this;
+                            _action();
+                        }
+                    }
+                    finally
+                    {
+                        if ((_state & (int)State.Started) != 0)
+                        {
+                            // Set the Finished flag to prevent a potential subsequent AbortThread call
+                            State oldState = (State)Interlocked.Or(ref _state, (int)State.Finished);
+
+                            if ((oldState & State.AbortRequested) != 0)
+                            {
+                                // Either in SFAR or SFA state
+                                while (true)
+                                {
+                                    // The enclosing finally may be cloned by the JIT for the non-exceptional code flow.
+                                    // In that case this code is not guarded against a thread abort, so make this FCall as
+                                    // soon as possible.
+                                    bool resetAbortRequest = ResetAbortThread();
+
+                                    // If there is an Abort in progress, we need to wait until it sets the TS_AbortRequested
+                                    // flag on this thread, then we can reset the flag and safely exit this frame.
+                                    if (((oldState & State.RunningAbort) == 0) || resetAbortRequest)
+                                    {
+                                        break;
+                                    }
+
+                                    // It should take very short time for AbortThread to set the TS_AbortRequested flag
+                                    Thread.Sleep(0);
+                                    oldState = (State)Volatile.Read(ref _state);
+                                }
+                            }
+
+                            t_execution = null;
```

Comment:
We set `t_execution` only after setting the `Started` flag, so we should be fine here.

> That said, it's not clear why the if condition is really needed, since _state should always be at least Started.

If `Abort` is called before we try to set the `Started` flag, then the `try` block does nothing; however, the `finally` block is still executed, that is why I needed the condition in the `finally` block.

Reply:
Right, that's only possible though because of where Register is being called.  If the call to Register were tightened to be just around the delegate invocation, such interleavings wouldn't be possible.  This relates to my suggested flow in https://github.com/dotnet/runtime/pull/71661#discussion_r927930923.